src/hardware/pwm.rs

```diff
@@ -0,0 +1,194 @@
+    lim_i_up2: Pt1<TIM4, C3>,
+    lim_i_up3: Pt1<TIM4, C4>,
+    lim_i_low0: Pt1<TIM3, C1>,
+    lim_i_low1: Pt1<TIM3, C2>,
+    lim_i_low2: Pt1<TIM3, C3>,
+    lim_i_low3: Pt1<TIM3, C4>,
+}
+
+impl Pwm {
+    /// Construct a new PWM driver for all Thermostat output channels.
+    ///
+    /// # Args
+    /// * `clocks` - Reference to CoreClocks
+    /// * `tim_rec` - Peripheral Reset and Enable Control for the timer 1/2/4
+    /// * `tim` - Timer 1/2/4 peripherals
+    /// * `pins` - PWM output pins
+    pub fn new(
+        clocks: &CoreClocks,
+        tim_rec: (rec::Tim1, rec::Tim3, rec::Tim4),
+        tim: (TIM1, TIM3, TIM4),
+        pins: PwmPins,
+    ) -> Pwm {
+        fn init_pwm_pin<P: PwmPin<Duty = u16>>(pin: &mut P) {
+            pin.set_duty(0);
+            pin.enable();
+        }
+
+        let (mut lim_v0, mut lim_v1, mut lim_v2, mut lim_v3) = tim.0.pwm(
+            (
+                pins.lim_v0_pin,
+                pins.lim_v1_pin,
+                pins.lim_v2_pin,
+                pins.lim_v3_pin,
+            ),
+            F_PWM,
+            tim_rec.0,
+            clocks,
+        );
+        init_pwm_pin(&mut lim_v0);
+        init_pwm_pin(&mut lim_v1);
+        init_pwm_pin(&mut lim_v2);
+        init_pwm_pin(&mut lim_v3);
+
+        let (mut lim_i_up0, mut lim_i_up1, mut lim_i_up2, mut lim_i_up3) = tim.2.pwm(
+            (
+                pins.lim_i_up0_pin,
+                pins.lim_i_up1_pin,
+                pins.lim_i_up2_pin,
+                pins.lim_i_up3_pin,
+            ),
+            F_PWM,
+            tim_rec.2,
+            clocks,
+        );
+        init_pwm_pin(&mut lim_i_up0);
+        init_pwm_pin(&mut lim_i_up1);
+        init_pwm_pin(&mut lim_i_up2);
+        init_pwm_pin(&mut lim_i_up3);
+
+        let (mut lim_i_low0, mut lim_i_low1, mut lim_i_low2, mut lim_i_low3) = tim.1.pwm(
+            (
+                pins.lim_i_low0_pin,
+                pins.lim_i_low1_pin,
+                pins.lim_i_low2_pin,
+                pins.lim_i_low3_pin,
+            ),
+            F_PWM,
+            tim_rec.1,
+            clocks,
+        );
+        init_pwm_pin(&mut lim_i_low0);
+        init_pwm_pin(&mut lim_i_low1);
+        init_pwm_pin(&mut lim_i_low2);
+        init_pwm_pin(&mut lim_i_low3);
+
+        Pwm {
+            lim_v0,
+            lim_v1,
+            lim_v2,
+            lim_v3,
+            lim_i_up0,
+            lim_i_up1,
+            lim_i_up2,
+            lim_i_up3,
+            lim_i_low0,
+            lim_i_low1,
+            lim_i_low2,
+            lim_i_low3,
+        }
+    }
+
+    /// Set PWM TEC limits.
+    ///
+    /// # Args
+    /// * `ch` - Thermostat output channel
+    /// * `limit` - TEC limit type
+    pub fn set(&mut self, ch: Channel, lim: Limit, val: f32) {
+        fn set_pwm<P: PwmPin<Duty = u16>>(pin: &mut P, duty: f32) {
+            let max = pin.get_max_duty();
+            let value = ((duty * (max as f32)) as u16).min(max);
```

Comment:
Why are you still clipping here? This was also commented on before and you marked it resolved.